### ISSUE TITLE
[css language server] Display color inside editor 

### ISSUE THREAD

**Issue by angelozerr** [mickaelistria/eclipse-bluesky#30]
The CSS Language server provides a custom `css/colorSymbols` request (see https://github.com/Microsoft/vscode/blob/master/extensions/css/server/src/cssServerMain.ts#L16) which provides a list of ranges where a color must be included in the editor:

I have implemented quickly this feature and here the result:

![image](https://user-images.githubusercontent.com/1932211/29542124-322ece96-86d9-11e7-9dab-292a87037846.png)

My main problem is that I have done that with reconciler and so it disable the HighlightReconciler from lsp4e because editor can have just one reconciler. So I'm waiting fix for https://bugs.eclipse.org/bugs/show_bug.cgi?id=520686 to contribute with this feature.

In other words I would like to have highlight symbol, folding and display color for the CSS editor. Thanks @mickaelistria  for your help!

**Comment by mickaelistria** [comment]
Is this red square inserted as a character in the editor (and text document)?

**Comment by angelozerr** [comment]
> Is this red square inserted as a character in the editor (and text document)?

No this red square is drawn with a `PaintObjectListener`.  When text editor content changes, it calls `css/colorSymbols` (with reconciler) which returns lsp4j range. This range is from start/end `red` text. After that I create a StyleRange with GlyphMetrics and paintObject display the red square. This red square is not editable. I'm using the same technic than http://www.java2s.com/Tutorial/Java/0280__SWT/StyledTextembedimages.htm

**Comment by angelozerr** [comment]
@mickaelistria If you wish I could push my work in my github project to see how it works.

**Comment by angelozerr** [comment]
Here a little demo:

![csscolordemo](https://user-images.githubusercontent.com/1932211/29580232-971dad94-8775-11e7-83cb-2b90711d4419.gif)

It's not perfect because the colorization is done with the character before the definition of the color. I don't know how to takes some place with StyledText without modifying the StyledText. I have tried with GlyphMetrics but it eats the first character of the color declaration.

If you are interested, tell me.

**Comment by angelozerr** [comment]
@mickaelistria have you interested with this feature. Please note that this color feature is a proposed feature  
https://github.com/Microsoft/vscode-languageserver-node/blob/9d6b53e0feb4039af95e2d108c376d8fc92d8b37/protocol/src/protocol.colorProvider.proposed.ts#L41

In other words I think lsp4e should host the coloration code with StyledText. I have seen (code) that JSON Server supports too color provider (if your JSON file is bound to a JSON Schema with "color" type. See https://github.com/Microsoft/vscode/blob/master/extensions/json/server/src/jsonServerMain.ts#L319 and https://github.com/Microsoft/vscode-json-languageservice/blob/bb6c48e833895d49f8174e4884356ba560b6dc60/src/test/documentSymbols.test.ts#L115

**Comment by mickaelistria** [comment]
Yes, this is a good proposal and I wish this makes it into BlueSky soon. I'm waiting for https://bugs.eclipse.org/bugs/show_bug.cgi?id=521418 to be fixed before we can properly integrate it in the Generic Editor.

**Comment by angelozerr** [comment]
> Yes, this is a good proposal and I wish this makes it into BlueSky soon. 

Cool!

I think we can do the same thing for JSON too. I'm waiting for https://bugs.eclipse.org/bugs/show_bug.cgi?id=521418  to do a PR. After you will able to move the code to lsp4e if you wish.

**Comment by angelozerr** [comment]
@mickaelistria I will soon do a PR. It works for CSS and JSON where CSS and JSON language server support it.

The feature that I don't know to implement is that color rectangle doesn't take place. You can do it by adding extra characters like http://www.java2s.com/Tutorial/Java/0280__SWT/StyledTextembedimages.htm but it update content of the editor.

Do you know if it's possible? Do you know where I could post my question? Create an eclipse bugzilla?

Thanks for your help!

**Comment by mickaelistria** [comment]
Adding space of non-payload character to a Text Editor is a hard thing.
This is already the topic of
https://bugs.eclipse.org/bugs/show_bug.cgi?id=507548 . I suggest you keep
discussing on this bugzilla and describe the new use-case there.

**Comment by mickaelistria** [comment]
Anything remaining on this issue or can we close it?

**Comment by angelozerr** [comment]
@mickaelistria I close it.
